bothan-core/src/registry/processor/weighted_median.rs

```diff
@@ -0,0 +1,217 @@
+            .collect();
+
+        let minimum_cumulative_weight = Decimal::deserialize(Decode::decode(decoder)?);
+
+        Ok(WeightedMedianProcessor {
+            source_weights,
+            minimum_cumulative_weight,
+        })
+    }
+}
+
+impl<'de> BorrowDecode<'de> for WeightedMedianProcessor {
+    fn borrow_decode<D: BorrowDecoder<'de>>(decoder: &mut D) -> Result<Self, DecodeError> {
+        let serialized_decimal_map: HashMap<String, [u8; 16]> =
+            BorrowDecode::borrow_decode(decoder)?;
+        let source_weights = serialized_decimal_map
+            .into_iter()
+            .map(|(k, v)| (k, Decimal::deserialize(v)))
+            .collect();
+
+        let minimum_cumulative_weight = Decimal::deserialize(BorrowDecode::borrow_decode(decoder)?);
+
+        Ok(WeightedMedianProcessor {
+            source_weights,
+            minimum_cumulative_weight,
+        })
+    }
+}
+
+impl WeightedMedianProcessor {
+    /// Creates a new `WeightedMedianProcessor`.
+    pub fn new(
+        source_weights: HashMap<String, Decimal>,
+        minimum_cumulative_weight: Decimal,
+    ) -> Self {
+        WeightedMedianProcessor {
+            source_weights,
+            minimum_cumulative_weight,
+        }
+    }
+}
+
+impl Process<(String, Decimal), Decimal> for WeightedMedianProcessor {
+    /// Processes the given data and returns the weighted median. If the cumulative weights of the
+    /// data sources are less than the minimum cumulative weight or the source associated
+    /// with the data does not have an assigned weight, it returns an error.
+    fn process(&self, data: Vec<(String, Decimal)>) -> Result<Decimal, ProcessError> {
+        let cumulative_weight = data.iter().try_fold(Decimal::zero(), |acc, (source, _)| {
+            self.source_weights
+                .get(source)
+                .map(|weight| acc + weight)
+                .ok_or(ProcessError::new(format!("Unknown source {source}")))
+        })?;
+
+        if cumulative_weight < self.minimum_cumulative_weight {
+            return Err(ProcessError::new(
+                "Not enough sources to calculate weighted median",
+            ));
+        }
+
+        let values = data
+            .into_iter()
+            .map(|(source, value)| {
+                self.source_weights
+                    .get(&source)
+                    .map(|weight| (value, *weight))
+                    .ok_or(ProcessError::new(format!("Unknown source {source}")))
+            })
+            .collect::<Result<Vec<(Decimal, Decimal)>, ProcessError>>()?;
+
+        let median = compute_weighted_median(values)
+            .ok_or_else(|| ProcessError::new("Failed to calculate weighted median"))?;
+        Ok(median)
+    }
+}
+
+fn compute_weighted_median<T>(mut values: Vec<(T, Decimal)>) -> Option<T>
+where
+    T: Ord + Copy + Add<Output = T> + Div<Output = T> + FromPrimitive,
+{
+    values.sort_by(|a, b| a.0.cmp(&b.0));
```

Comment:
or we just `values.sort_by_key(|k| k.0);`

Reply:
sort_by_key will probably requires copy to return type K, will remove copy trait (as its unnecessary) and keep as sort_by but with tuple unpacking